Support Configuration for Cipher Suits and Protocols used by Upstream and Downstream

alanrobinson-dwp opened the pull request:
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

Attempt at allowing users to pass Cipher Suites and Enabled Protocols in TLS Configuration so they are set when the SSLContext is built.

### Additional Context

This is to address https://github.com/kroxylicious/kroxylicious/issues/1006

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [ ] Update documentation
- [ ] Make sure all unit/integration tests pass
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [ ] Ensure the PR references relevant issue(s) so they are closed on merging.
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

alanrobinson-dwp commented:
Hi guys, please find a first draft at passing through Cipher Suites and Enabling Protocols for issue https://github.com/kroxylicious/kroxylicious/issues/1006. Hopefully it's close to what you had in mind.

I still have a couple of areas I'd like to address which I'm having a frustrating time thinking on best way to implement. So ideas and comments will be much appreciated

- Validating the configuration. Ideally I'd like to validate the list of ciphers and protocols against a valid list, but there doesn't seem to be an obvious place to get this. Currently any list of Strings can be passed, and if there are incorrect values, you see Invalid SNI errors when communication takes place.
- Integration Tests. I was hoping to write one that would set client up to use a certain Cipher and broker to accept only that, but couldn't see an obvious way of doing that. I've had a few iterations but haven't put anything into PR yet. At the moment I can pass a correct Cipher and/or Protocol into the ConfigurationBuilder (e.g. withCipherSuites) and an existing IT works, and fails if it's a duff Cipher, but I can't prove it's working upstream by setting them both end.

Thanks
Alan

k-wall commented:
@alanrobinson-dwp thank you for the contribution, I'll get to the review as soon as I can

k-wall commented:
I am thinking about the use-case the security researchers find a security flaw in a established cipher suite.  The industry reacts to the news by updating configurations to remove the ciphersuite from production servers.  We want to make this as painless as possible.

Whilst your suggested API supports this use-case, it isn't terribly convenient.  I'm forced to list all of the ciphersuites I want to **keep**.   That's tricky - it might depend on the version of Java that in use.  Also it doesn't convey the intent very well.   The config will probably rot over time too as ciphersuites come and go.

With this in mind, I'd like to suggest a slightly different API. 

```yaml
tls:
  enabledCipherSuites:                     # list of cipher suites to be enabled, in priority order. if null, the platform default cipher suites will be used.
  - TLS_AES_256_GCM_SHA384
  - etc
  disabledCipherSuites:                     # list of the cipher suite(s) to be disabled.  if null, no cipher suites will be disabled.
  - TLS_CHACHA20_POLY1305_SHA256
  enabledProtocols:                           # list of TLS protocols enabled. if null, the platform default protocols will be used.
  - TLSv1.3
 disabledProtocols:                           #  list of the protocols to be disabled.  if null, no protocols will be disabled.
  - TLSv1.2
```

So in the use-case I've outlined above, the user would just need something like:

```yaml
tls:
  disabledCipherSuites:
  - TLS_CHACHA20_POLY1305_SHA256
```

which is much easier to read.

If you use case is just to limit your services to a select set of cipher suites to comply with a local policy, you can do that too.

```yaml
  enabledCipherSuites: 
  - TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
  - TLS_DHE_DSS_WITH_AES_256_GCM_SHA384
  - TLS_DHE_RSA_WITH_AES_256_GCM_SHA384
```

@alanrobinson-dwp @ddaviesbrackett-dwp @kroxylicious/developers  WDYT?

SamBarker commented:
> I am thinking about the use-case the security researchers find a security flaw in a established cipher suite. The industry reacts to the news by updating configurations to remove the ciphersuite from production servers. We want to make this as painless as possible.

My philosophy when thinking about API design (and I consider config models to be part of API design) is to make the common case easy and as much of everything else possible. 

I'm not quite sure that setting the bar at evolving a cipher list reaches the common case, but I agree it should be possible.

> 
> With this in mind, I'd like to suggest a slightly different API.
> 
> ```yaml
> tls:
>   enabledCipherSuites:                     # list of cipher suites to be enabled, in priority order. if null, the platform default cipher suites will be used.
>   - TLS_AES_256_GCM_SHA384
>   - etc
>   disabledCipherSuites:                     # list of the cipher suite(s) to be disabled.  if null, no cipher suites will be disabled.
>   - TLS_CHACHA20_POLY1305_SHA256
>   enabledProtocols:                           # list of TLS protocols enabled. if null, the platform default protocols will be used.
>   - TLSv1.3
>  disabledProtocols:                           #  list of the protocols to be disabled.  if null, no protocols will be disabled.
>   - TLSv1.2
> ```

I'm not a fan of giving `null` meaning at the best of times but especially when it comes to security properties I think we should be explicit and model `use platform [cipher suites|protocols]` explicitly. 

> So in the use-case I've outlined above, the user would just need something like:
> 
> ```yaml
> tls:
>   disabledCipherSuites:
>   - TLS_CHACHA20_POLY1305_SHA256
> ```
> 
> which is much easier to read.
> 
> If you use case is just to limit your services to a select set of cipher suites to comply with a local policy, you can do that too.

Do you see `enabled` and `disabled` being used at the same time? or is it an either or? If so I wonder if the nodes should be `onlyEnableCipherSuites` or `allowOnlyCipherSuites`. 

My biggest concern with this model is it gets very verbose especially when multiple virtual clusters are involved. Users start having to ensure they update the TLS nodes across multiple virtual clusters. The trouble is the solution to this leads in to the previously debated and unresolved question of config defaults and scopes. Is configuring TLS proxy wide a tenable position? Would we need the potential to [override|extend] (which?) the config at the virtual cluster level?

What about the admin http server? Does that have its own set of ciphers and protocols to deal with (internal only interface being less stringent than public facing?)

I think thats a very long non answer to @k-wall's question. So in more concrete terms. 

```
tls:
  cipherSuites:
    allowed:
       PLATFORM
    denied:     
       - TLS_CHACHA20_POLY1305_SHA256
  protocols:
     allowed:
        PLATFORM
     denied:
         - TLSv1.2        
```
Where allowed conforms to
```json
        "oneOf": [
          {
            "type": "string",
            "pattern": "PLATFORM"
          },
          {
            "type": "array",
            "items": [
              {
                "type": "string"
              }
            ]
          }
        ]
```
I think we also have to explicitly say that entires in deny take precedence over entries in allow/enable. If something exists in the deny list it has to be denied. 

My hunch is that most users will just want to inherit the defaults for their JVM but I think we should make that explicit.

k-wall commented:
@SamBarker I like your suggestion to use objects for `cipherSuites` and `protocols`.   I'm less convinced by the need to for `PLATFORM` to signify use-defaults.   The use of `null` to mean "use the default" is so well established I'd argue it is least surprise.  I also value the succinctness.

> Do you see enabled and disabled being used at the same time? or is it an either or? If so I wonder if the nodes should be onlyEnableCipherSuites or allowOnlyCipherSuites.

In practice it will be either/or, but I wouldn't disallow both.

```
tls:
  cipherSuites:
    allowed:
       - A
       - B
       - C
    denied:     
       - B
```

means cipher suites A, C will be enabled in the SSL Engine.


> My biggest concern with this model is it gets very verbose especially when multiple virtual clusters are involved. 

I agree with that, but I don't think we need to conflate that with this PR.    You should be able to default the TLS settings higher up (virtual cluster? globally?) and still have the ability to override on a case by case basis.  

That allows a use-case where, say an organisation has a policy where is wants to use TLS1.3 everywhere but gives an escape hatch that allows the proxy to still talk to that dusty old KMS in the corner that only supports TLS 1.2.

> What about the admin http server?

We've got a metrics endpoint right now rather than a admin.   It is uses plaintext HTTP.

If we had an admin server, I would expect we'd have the same protocol/cipher suite controls.

We will want to apply the ciphersuite and protocol controls to the HTTPS clients used by the AWS and Vault KMSs and the Schema Registry.   If be happy to see these come as follow up PRs.

alanrobinson-dwp commented:
thanks guys, I'll pick these comments up and make some amendments

SamBarker commented:
>I agree with that, but I don't think we need to conflate that with this PR. You should be able to default the TLS settings higher up (virtual cluster? globally?) and still have the ability to override on a case by case basis.

Yes, sorry I should have been clearer I wasn't trying to make that part of this PR. There is still a lot of thinking to do about how to make configuration inheritance work sensibly.

> The use of null to mean "use the default" is so well established I'd argue it is least surprise. I also value the succinctness.

I agree the pattern is common and succinct (which is a good property to have) but when I'm browsing config for a project I'm not intimately familiar with I've often found myself asking the question "but what does it do if I leave it blank/empty?" "what are the defaults if I leave it blank?". So I'm suggesting we follow Tim Peters advice from the [Zen of python](https://peps.python.org/pep-0020/) `Explicit is better than implicit` and `Readability counts`. 

All that said `PLATFORM` vs `null` is not a hill I'm going to die on. If the general view is `null` works then lets do it.

> In practice it will be either/or, but I wouldn't disallow both.

Yeah, fair enough. It supports pasting in the standard set of `Allowed` and a custom override to Deny it in the specific context.

>We've got a metrics endpoint right now rather than a admin. It is uses plaintext HTTP.

Fair, I don't think its a specialist concern, it can just grow a TLS node when needed.

alanrobinson-dwp commented:
Hi @k-wall @SamBarker I've made some amendments from your comments and created a CipherSuiteFilter (let me know what you think and if it's in the right place?) that will allow us to Deny specific CipherSuites and the configuration input will now match what you were describing, but let me know if I've gone down the right track.

I've not yet implemented the idea of the protocols allowed and denied as there isn't a similar Interface like with the CipherSuites so we'd probably have to do something custom within the code. Currently for protocols it'll just pass in what we want to allow. If you're happy with the cipherSuites I can look to do something similar but probably custom on the Kroxy side

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,56 @@
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.filter;
```
`CipherSuiteFilter` a filter but not _that_ kind of filter! The filter package is for the machinery used for Kroxylicious Filters, so it is not appropriate place for the new class.

The class is used exclsuively by the VirtualCluster class, so I would make it a sibling of that class. In other words, put it in `io.kroxylicious.proxy.model` (and make it package private).  The fact that `VirtualCluster` is in package called `io.kroxylicious.proxy.model` is weird but let's not worry about that here!

alanrobinson-dwp replied:
No worries, I'll get it moved. I knew it was probably in the wrong place when I was scanning the directories deciding where to put it

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,56 @@
+ * If ciphers are declared then they are used instead of the default platform ciphers
+ * Both lists would then have anything removed that wasn't supported or in the deniedCiphers
+ */
+public final class DenyCipherSuiteFilter implements CipherSuiteFilter {
```
The name does really reflect the responsibility of the class.  It is the lone word 'Deny' that seems wrong.  It is responsible for both allow and deny.   I would suggest `ConfigurableCipherSuiteFilter`.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public final class DenyCipherSuiteFilter implements CipherSuiteFilter {
+
+    public static final DenyCipherSuiteFilter INSTANCE = new DenyCipherSuiteFilter();
```
The singleton idea doesn't really work.   There might be many VirtualClusters defined each with different cipher suite requirements.   Which one's rules would win?   They'd all call `io.kroxylicious.proxy.filter.DenyCipherSuiteFilter#deniedCiphers`.  It would be a race.

Get rid of the singleton and have let `VirtualCluster` construct an instance of ConfigurableCipherSuiteFilter, passing its own configuration.  Pass the constructed instance to the io.netty.handler.ssl.SslContextBuilder#ciphers(java.lang.Iterable<java.lang.String>, io.netty.handler.ssl.CipherSuiteFilter)

You probably want a ctor like this:

```java
ConfigurableCipherSuiteFilter(@Nullable List<String> allowed, @Nullable List<String> denied) {}

```

alanrobinson-dwp replied:
Yes, good shout! I was using the 2 implemented Netty classes Identity* and Supported* as templates which I why I went down that route. Hadn't even thought. Will sort out

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public String[] filterCipherSuites(Iterable<String> ciphers, List<String> defaultCiphers,
+                                       Set<String> supportedCiphers) {
+        return ciphers == null ? allowedCiphers(defaultCiphers.size(), defaultCiphers, supportedCiphers)
```
you can probably apply some Java streams goodness(TM).  

```java
        var actualCiphers = (ciphers == null || !ciphers.iterator().hasNext()) ? defaultCiphers : ciphers;

        return StreamSupport.stream(actualCiphers.spliterator(), true)
                .filter(supportedCiphers::contains)
                .filter(Predicate.not(c -> deniedCiphers.contains(c)))
                .toList()
                .toArray(new String[] {});
```

k-wall replied:
Also I think we should log (warn) if either ciphers or deniedCiphers contain a cipher suite name that's not  in the supportedCiphers set.   It could be that a user has typo'd the name of teh cipher suite they are trying to exclude.

alanrobinson-dwp replied:
Yes, I did have some logging in when doing the development actually, but stripped it after finishing. Will add in the warn for things that may be wrong.

I can definitely add some streams goodness haha. As mentioned above I'd taken inspiration from the Netty Identity* and Supported* implemented classes which suited my old dog/no new tricks coding!

alanrobinson-dwp replied:
The code you have above, could the parallel true potentially risk the order of the list passed in? I believe the ciphers are in order of preference

SamBarker replied:
I doubt we should be enabling parallel streams in this case, the lists are going to be so short the costs of going parallel are not worth it.

k-wall commented on kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilterTest.java:
```diff
@@ -0,0 +1,79 @@
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+
+public class DenyCipherSuiteFilterTest {
```
```suggestion
class DenyCipherSuiteFilterTest {
```

k-wall commented on kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilterTest.java:
```diff
@@ -0,0 +1,79 @@
+public class DenyCipherSuiteFilterTest {
+
+    @Test
+    public void emptyCiphersDefaultsToDefaultCiphers() {
```
```suggestion
   void emptyCiphersDefaultsToDefaultCiphers() {
```

k-wall commented on kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilterTest.java:
```diff
@@ -0,0 +1,79 @@
+
+        String[] filtered = DenyCipherSuiteFilter.INSTANCE.deniedCiphers(null)
+                .filterCipherSuites(null, defaultCiphers, supportedCiphers);
+        assertArrayEquals(defaultAllowedCiphers.toArray(), filtered);
```
we are favouring assertj for all new word.. it is just much more expressive and concise.

```suggestion
        assertThat(filtered).containsAll(defaultAllowedCiphers);
```

k-wall commented:
Nice work @alanrobinson-dwp , looks like it is heading in a good direction.  I've left some review feedback to be considered.

robobario commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,56 @@
+        List<String> newCiphers = new ArrayList<>(newListSize);
+        for (String cipher : ciphersToUse) {
+            if (cipher == null) {
+                break;
```
when would we expect a null value to be contained in the `ciphersToUse`? maybe the `AllowDeny` should disallow null items so that:

```yaml
cipherSuites:
  allow: null
  deny: null
```
is valid, but:
```yaml
cipherSuites:
  allow:
  - null
  - 'cipher'
  deny:
  - null
  - 'cipher2'
```
is invalid and this code should either skip the null or fail with explosion rather than breaking out of the loop.

alanrobinson-dwp replied:
Hi @robobario I'd been using the Netty implemented classes like [SupportedCipherSuiteFilter](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SupportedCipherSuiteFilter.java) for what I did and they all seemed to have that check, but you're absolutely right. It shouldn't be needed. 

I'll look at the validation of how these objects will be generated

robobario replied:
well, that's ominous

SamBarker replied:
If I'm reading https://github.com/netty/netty/commit/04f77b76f802a5bb30447eb0e3580f6ba0e1f2db correctly the Netty impl validates earlier in the process so the break on `null` should never actually happen. That said I think @robobario is right we should continue rather than break.

robobario commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/filter/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,56 @@
+    private DenyCipherSuiteFilter() {
+    }
+
+    private List<String> deniedCiphers;
```
Should be a `Set` I think

k-wall replied:
`deniedCiphers` is a set.  The order of the denies doesn't matter.

`allowedCiphers` has to be ordered, as you say.     The best choice would be a `SequencedSet`, but we are not on Java 21 and SequencedSet wont play nicely with direction #1639 is going.   Modelling it as a List<String> will be fine. Your DenyCipherSuiteFilters#ciphers method impl will receive the allowedCiphers asthe  Iterable ciphers arg. Order will be maintained.

alanrobinson-dwp replied:
Ah yes of course, denied doesn't matter, apologies @robobario you were right

SamBarker reviewed (commented):
Thanks for pushing this forward @alanrobinson-dwp

alanrobinson-dwp commented on kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:
```diff
@@ -16,7 +17,33 @@
+     * {@see io.kroxylicious.proxy.config.tls.Tls#Tls(
+     *      io.kroxylicious.proxy.config.tls.KeyProvider,
+     *      io.kroxylicious.proxy.config.tls.TrustProvider,
+     *      List<java.lang.String>,
+     *      List<java.lang.String>)
```
Good spot

alanrobinson-dwp commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/TlsIT.java:
```diff
@@ -352,9 +353,51 @@ void downstreamAndUpstreamTls(Class<? extends PasswordProvider> providerClazz, @
     }
 
+    @Test
+    void downstreamMutualTls_SuccessfulTlsClientAuthRequiredWithCipherSuitesAllowed(KafkaCluster cluster) throws Exception {
```
Yeah that's a fair call. I was worried about too much boiler plate which is why I tried to reuse the same config builder, but it's not great. I'll get them split out and add in one to test the denied

SamBarker replied:
Yeah, I hear you! 

For the record I've got a much higher tolerance for repeated boiler plate in tests, as it it often makes it clearer what is being asserted, though I will often resort to test specific factory type methods to hide the boiler plate. I really want **a reader** to understand the 
> given this X input when I do Y we get Z

I also don't like being overly restrictive on what `Z` is. I want tests that have **a single** reason to fail! All this means that test run time and verbosity are very much a 2nd order concerns for me.

alanrobinson-dwp commented on kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:
```diff
@@ -16,7 +17,33 @@
+    public Tls(KeyProvider key, TrustProvider trust, AllowDeny<String> cipherSuites, AllowDeny<String> protocols) {
+        this.key = key;
+        this.trust = trust;
+        this.cipherSuites = cipherSuites;
+        this.protocols = protocols;
```
Thanks @SamBarker, are you thinking that the configuration would look something like

```
key:
    ...
trust:
    trustOptions:
        clientAuth: REQUIRED
        protocols:
            allowed:
              - TLSv1.3
            denied
              - TLSv1.2
cipherSuites:
    allowed:
       - CIPHERS
    denied
       - CIPHERS
```

k-wall replied:
I don't think @SamBarker is right.

Protocol inclusion/exclusion applies equally to both the TLS client and server peers.  https://datatracker.ietf.org/doc/html/rfc8446 describes how the client uses supported_versions extension to describe to the server which TLS version(s( it supports, which is used for protocol negotiation.  What's more Netty's SSLContextBuilder and the JDK's SSLEngine treat TLS protocol symmetrically. 

I think `tls` is the right place for `protocols`.

I notice Kafka's client side has https://kafka.apache.org/documentation/#brokerconfigs_ssl.protocol to tell the client to use a particular TLS protocol (even if other TLS protocol versions are enabled).  However, I'm not convinced we need an analogue of this configuration.  If someone wants to lock the client to a particular version, the allow/deny config allows that to be expressed.

SamBarker replied:
Happy is the wrong term content to be wrong would be more accurate. 

That said I did learn something here so not all is lost.

alanrobinson-dwp commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +256,33 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+        var protocols = Optional.ofNullable(tlsConfiguration.protocols());
+
+        if (protocols.isPresent()) {
+            var allowedProtocols = Optional.ofNullable(protocols.get().allowed()).orElse(Set.of(Protocols.TLSv1_2, Protocols.TLSv1_3));
```
I'm in the middle of getting a Platforms supported by Default protocols to fill this by the way, but thought best to commit in the rest of changes in case other comments.

alanrobinson-dwp commented:
Hi @k-wall @SamBarker @robobario 

I've added some changes in that hopefully address a lot of the points raised. I've also added some stuff in to try and control the logic around determining protocols. I'd be interested in your views. I have used an Enum to drive the validation on what you can enter, and then use platforms supported protocols if someone happened to provide a denied protocol and no allowed. The risk is the Enum doesn't have one, but I think after some reading I've got them covered

k-wall commented on kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/AllowDeny.java:
```diff
@@ -0,0 +1,19 @@
+ * @param allowed specifies a list of allowed objects.
+ * @param denied specifies a list of denied objects.
+ */
+public record AllowDeny<T>(@Nullable Set<T> allowed, @Nullable Set<T> denied) {}
```
`allowed` requires order, so we should its declared type should be a List.

k-wall commented on kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/SslProtocol.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Arrays;
+import java.util.Optional;
+
+public enum SslProtocol {
```
We favoured using the word Tls over Ssl in the code, so let's call this TlsProtocol.

k-wall commented on kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/SslProtocol.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Optional;
+
+public enum SslProtocol {
+    SSLv2("SSLv2Hello"),
```
Can you make the enums comply with [java:S115](https://next.sonarqube.com/sonarqube/coding_rules?open=java%3AS115&rule_key=java%3AS115)?

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,45 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+            var protocols = Optional.ofNullable(tlsConfiguration.protocols());
+
+            if (protocols.isPresent()) {
+                var allowedProtocols = Optional.ofNullable(protocols.get().allowed())
+                        .orElse(Arrays.stream(SSLContext.getDefault().getSupportedSSLParameters().getProtocols())
+                                .map(sslProtocol -> SslProtocol.getProtocolName(sslProtocol).get()).collect(Collectors.toSet()));
```
```suggestion
                var allowedProtocols = Optional.ofNullable(protocols.get().allowed())
                        .orElse(Arrays.stream(SSLContext.getDefault().getSupportedSSLParameters().getProtocols())
                                .map(SslProtocol::getProtocolName)
                                .filter(Optional::isPresent)
                                .map(Optional::get)
                                .collect(Collectors.toSet()));
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,45 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+                        .orElse(Arrays.stream(SSLContext.getDefault().getSupportedSSLParameters().getProtocols())
+                                .map(sslProtocol -> SslProtocol.getProtocolName(sslProtocol).get()).collect(Collectors.toSet()));
+                var deniedProtocols = Optional.ofNullable(protocols.get().denied())
+                        .orElse(Collections.EMPTY_SET);
```
```suggestion
                        .orElse(Set.of());
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,45 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+                    sslContextBuilder.protocols(protocolsToUse);
+                }
+                else {
+                    LOGGER.warn("The protocols configuration you have in place has resulted in platform defaults being used");
```
I think we should fail here rather than just logging a warning.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,45 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+                        .orElse(Collections.EMPTY_SET);
+
+                var protocolsToUse = allowedProtocols.stream()
+                        .filter(Predicate.not(p -> deniedProtocols.contains(p)))
```
```suggestion
                        .filter(Predicate.not(deniedProtocols::contains))
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,45 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+                var protocolsToUse = allowedProtocols.stream()
+                        .filter(Predicate.not(p -> deniedProtocols.contains(p)))
+                        .map(SslProtocol::getSslProtocol)
+                        .collect(Collectors.toList());
```
```suggestion
                        .toList();
```

k-wall commented on kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/SslProtocol.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Arrays;
+import java.util.Optional;
+
+public enum SslProtocol {
```
I'm not convinced that introducing this enum is the right thing to do.   Both the JDK and the Netty treat TLS Protocols name as a String.  I think we should do the same.   It is possible alternative TLS provider implementation name things in a different way.

k-wall replied:
Use SSLContext.getDefault().getSupportedSSLParameters().getProtocols() to validate that the protocol names  in the allow and deny lists are known to the platform.   Log a warning if they are not. IOW do what you did for ciphersuites.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,45 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+
+            if (protocols.isPresent()) {
+                var allowedProtocols = Optional.ofNullable(protocols.get().allowed())
+                        .orElse(Arrays.stream(SSLContext.getDefault().getSupportedSSLParameters().getProtocols())
```
Using `SSLContext.getDefault().getSupportedSSLParameters().getProtocols()` to get the default protocols only works if the user is using the JDK for TLS.    If the user has chosen to use the OpenSSL, then it is going to give the wrong answer.  See `io.netty.handler.ssl.OpenSsl#defaultProtocols(boolean)`.

That said, I think what you have is good enough for this PR. 

It is a pity Netty doesn't have a builder method (@franz1981 WDYT?)
```
SslContextBuilder protocols(Iterable<String> protocols, ProtocolFilter protocolFilter)
```

k-wall commented on kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/Tls.java:
```diff
@@ -16,7 +16,33 @@
 public record Tls(KeyProvider key,
-                  TrustProvider trust) {
+                  TrustProvider trust,
+                  AllowDeny<String> cipherSuites,
```
Can we add Javadoc for the new parameters?

k-wall commented on kroxylicious-api/src/test/java/io/kroxylicious/proxy/config/tls/TlsParseTest.java:
```diff
@@ -439,7 +442,62 @@ void testKeyStoreFilePathStorePasswordProvider() throws IOException {
+
+    @Test
+    void testEnabledProtocolsProvider() throws IOException {
+        SslProtocol tst = SslProtocol.getProtocolName("TLSv1.2").get();
```
redundant line?

alanrobinson-dwp replied:
Not sure how that snuck in there! Cheers

k-wall commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/TlsIT.java:
```diff
@@ -352,6 +355,234 @@ void downstreamAndUpstreamTls(Class<? extends PasswordProvider> providerClazz, @
+    void downstreamMutualTls_SuccessfulTlsWithProtocolsAllowed(KafkaCluster cluster) throws Exception {
+        var bootstrapServers = cluster.getBootstrapServers();
+
+        // Cipher we want to use
```
```suggestion
        // Protocol we want to use
```

k-wall commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/TlsIT.java:
```diff
@@ -352,6 +355,234 @@ void downstreamAndUpstreamTls(Class<? extends PasswordProvider> providerClazz, @
     }
 
+    @Test
+    void downstreamMutualTls_SuccessfulTlsWithProtocolsAllowed(KafkaCluster cluster) throws Exception {
```
You've named the test "mutual", but I don't think we actually need mutual authentication in this test.   I'd get rid of the word "mutual" from the new tests, and remove the  client's `SSL_KEYSTORE` configuration.

k-wall commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/TlsIT.java:
```diff
@@ -352,6 +355,234 @@ void downstreamAndUpstreamTls(Class<? extends PasswordProvider> providerClazz, @
+            assertThat(createTopicsResult.all()).isDone();
+        }
+    }
+
```
Could we have a ITs that verify that the new options work on the upstream side too?

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,57 @@
+        this.deniedCiphers = deniedCiphers == null ? new HashSet<>() : deniedCiphers;
+    }
+
+    private Set<String> deniedCiphers;
```
```suggestion
    private final Set<String> deniedCiphers;
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,57 @@
+
+        StreamSupport.stream(actualCiphers.spliterator(), false)
+                .filter(Predicate.not(actualCipher -> supportedCiphers.contains(actualCipher)))
+                .forEach(unsupportedCipher -> LOGGER.warn("Cipher {}, not in Supported Ciphers", unsupportedCipher));
```
```suggestion
                .forEach(unsupportedCipher -> LOGGER.warn("Ignoring allowed cipher '{}' as it is not recognized by this platform (supported ciphers: {})", unsupportedCipher, supportedCiphers));
```

k-wall commented on kroxylicious-runtime/src/test/java/io/kroxylicious/proxy/model/DenyCipherSuiteFilterTest.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Test
+    void emptyCiphersDefaultsToDefaultCiphers() {
+        List<String> defaultCiphers = Arrays.asList("CIPHER_SUITE_1", "CIPHER_SUITE_2", "CIPHER_SUITE_3", "CIPHER_SUITE_4");
```
Does the test data make sense? I would expect that every member of the `defaultCiphers` would always be a member of `supportedCiphers`.  In other words this:

```
        List<String> defaultCiphers = List.of("CIPHER_SUITE_1", "CIPHER_SUITE_2", "CIPHER_SUITE_3");
        Set<String> supportedCiphers = Set.of("CIPHER_SUITE_1", "CIPHER_SUITE_2", "CIPHER_SUITE_3", "CIPHER_SUITE_4");
        List<String> defaultAllowedCiphers = List.of("CIPHER_SUITE_1", "CIPHER_SUITE_2", "CIPHER_SUITE_3");
```

alanrobinson-dwp replied:
That's a good question, the defaultCiphers and supportedCiphers get determined within the SslContextBuilder. I'll take a look into the code to find out what they can be. 

I'd focused on testing the functionality I'd created within the custom DenyCipherSuiteFilter was doing what I expected and not really considered whether the data was realistic. So in this test I was wanting to make sure the supported ciphers logic was working.

Let me get into the SslContextBuilder code and double check my understanding of "default" and "supported" ciphers.

alanrobinson-dwp replied:
I've had a dig into Netty code (https://github.com/netty/netty/blob/9f0b38b7b6e01638075aa981872be306532f804b/handler/src/main/java/io/netty/handler/ssl/JdkSslContext.java#L111C13-L111C29) and your is correct so I've updated the test to be a bit tidier and take this into account

k-wall commented:
@alanrobinson-dwp HNY.  Thanks for the update to the PR.  It is looking better.  I've left some comments to be considered.

alanrobinson-dwp commented:
> @alanrobinson-dwp HNY. Thanks for the update to the PR. It is looking better. I've left some comments to be considered.

Happy New Year! Thanks for the comments. I'll get onto these today/tomorrow

alanrobinson-dwp commented:
Hi @k-wall, I've made changes based on your comments. The only outstanding changes would be based on stripping the Enum out back to just a list of Strings and passing that way, which I've added a comment to, and also the comment around `SSLContext.getDefault().getSupportedSSLParameters().getProtocols()` and whether this okay as is.

Let me know what you think

Cheers
Alan

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/DenyCipherSuiteFilter.java:
```diff
@@ -0,0 +1,58 @@
+
+        StreamSupport.stream(deniedCiphers.spliterator(), false)
+                .filter(Predicate.not(deniedCipher -> supportedCiphers.contains(deniedCipher)))
+                .forEach(unsupportedCipher -> LOGGER.warn("Denied Cipher {}, not in Supported Ciphers", unsupportedCipher));
```
```suggestion
                .forEach(unsupportedCipher -> LOGGER.warn("Ignoring denied cipher '{}' as it is not recognized by this platform (supported ciphers: {}", unsupportedCipher));
```

k-wall replied:
We want it to be clear in the logs that the we are ignoring the unrecognised item.

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -79,23 +85,41 @@ private static void logVirtualClusterSummary(String clusterName, TargetCluster t
         }
     }
 
+    public static String generateTlsSummary(Optional<Tls> tlsToSummarize) {
```
```suggestion
    private static String generateTlsSummary(Optional<Tls> tlsToSummarize) {
```

k-wall replied:
nit

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,50 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+                                .map(TlsProtocol::getProtocolName)
+                                .filter(Optional::isPresent)
+                                .map(Optional::get)
+                                .collect(Collectors.toList()));
```
```suggestion
                                .toList());
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,50 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+                    sslContextBuilder.protocols(protocolsToUse);
+                }
+                else {
+                    throw new IllegalStateException(
```
```suggestion
                    throw new IllegalConfigurationException(
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,50 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
+
+            if (protocols.isPresent()) {
+                var allowedProtocols = Optional.ofNullable(protocols.get().allowed())
+                        .orElse(Arrays.stream(SSLContext.getDefault().getSupportedSSLParameters().getProtocols())
```
I _think_ we want to base on the default ssl parameters rather than the platform 
rather than the supported.  The platform might support an aged `SSLv1`, but it is turned off by default.  We don't want to inadvertently enable historic stuff.



```suggestion
                        .orElse(Arrays.stream(SSLContext.getDefault().getDefaultSSLParameters().getProtocols())
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/model/VirtualCluster.java:
```diff
@@ -229,6 +259,50 @@ private static NettyTrustProvider configureTrustProvider(Tls tlsConfiguration) {
 
+    private static void configureCipherSuites(SslContextBuilder sslContextBuilder, Tls tlsConfiguration) {
+        Optional.ofNullable(tlsConfiguration.cipherSuites())
+                .map(ciphers -> sslContextBuilder.ciphers(
```
nit:  you are relying on map having a side effect.  that's not how map is intended to be used.  `ifPresent` communicates the intent to the reader.

```suggestion
                .ifPresent(ciphers -> sslContextBuilder.ciphers(
```

k-wall commented:
@alanrobinson-dwp : I made some suggestions in https://github.com/alanrobinson-dwp/kroxylicious/pull/1/

k-wall commented on kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/TlsProtocol.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.Arrays;
+import java.util.Optional;
+
+public enum TlsProtocol {
```
I'm still in favour of getting rid of this enum for the reasons I expressed before (both the JDK and Netty treat the protocol as Strings, so we should too).

alanrobinson-dwp replied:
Thanks @k-wall, yeah I'll get rid of that. I'll update the other bits first then strip out the Enum

k-wall commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/TlsIT.java:
```diff
@@ -352,6 +355,374 @@ void downstreamAndUpstreamTls(Class<? extends PasswordProvider> providerClazz, @
+            // do some work to ensure connection is opened
+            final CreateTopicsResult createTopicsResult = createTopic(admin, TOPIC, 1);
+            assertThat(createTopicsResult.all()).isDone();
+        }
```
We could consider strengthening the test by actually asserting that the client is truly using the protocol we intend.   I know you'd normally avoid breaking the client's abstraction in this way, but as this is security relevant, I think it is worth considering. 

Assert-J does a good job of hiding the reflection vomit that would otherwise be required.

```suggestion
            // verify that the admin is actually using the intended protocol
            assertThat(admin)
                    .extracting("instance")
                    .extracting("client")
                    .extracting("selector")
                    .extracting("channels", InstanceOfAssertFactories.map(String.class, Object.class))
                    .anySatisfy((id, channel) -> {
                        assertThat(channel)
                                .extracting("transportLayer")
                                .extracting("sslEngine", InstanceOfAssertFactories.type(SSLEngine.class))
                                .extracting(SSLEngine::getSession)
                                .extracting(SSLSession::getProtocol)
                                .isEqualTo("TLSv1.2");
                    });
        }
```

The same thing works for cipher suite- javax.net.ssl.SSLSession#getCipherSuite

k-wall commented:
@alanrobinson-dwp just want to make sure you saw: https://github.com/alanrobinson-dwp/kroxylicious/pull/1

alanrobinson-dwp commented:
> @alanrobinson-dwp just want to make sure you saw: [alanrobinson-dwp#1](https://github.com/alanrobinson-dwp/kroxylicious/pull/1)

Hi @k-wall, yes, just looking at it now. I'd made some changes yesterday from your previous comment, set away a maven build before I left to check I'd not broken anything, then committed and pushed it this morning before checking my mail and seeing the PR.

I'm just tallying off what comments I'd already actioned with what's in the PR. Looks good so far, I'll start incorporating as I go through the changes. I definitely need to brush up on my method referencing!

k-wall commented:
I just rebased the PR based on your fork

alanrobinson-dwp commented:
> I just rebased the PR based on your fork

Great thanks. I'll pull it in and work from there, and get the Enum stripped out

k-wall commented:
> Great thanks. I'll pull it in and work from there, and get the Enum stripped out

I forgot to add the DCO onto my commit.   I can either sort that out when it comes to merging this work, or if you want, just squash all the commits together.    Whichever you prefer is fine.

alanrobinson-dwp commented:
> > Great thanks. I'll pull it in and work from there, and get the Enum stripped out
> 
> I forgot to add the DCO onto my commit. I can either sort that out when it comes to merging this work, or if you want, just squash all the commits together. Whichever you prefer is fine.

Hi @k-wall, I've squashed the commits and I think I've addressed all the comments. Please let me know if I've missed something or misunderstood on anything! Thanks

k-wall commented:
@alanrobinson-dwp can you add a CHANGELOG entry and add yourself to the CONTRIBUTORS file (if you like).  That's for the great work.

k-wall commented:
<img width="1015" alt="Screenshot 2025-01-16 at 15 50 07" src="https://github.com/user-attachments/assets/d39d80fd-6c09-4b25-806d-913efb10dcf8" />
